Relax mutation to aten.to in export

yushangdi opened the pull request:
Summary:
Today, we force aten.to to always copy under the hood and ban the mutation on the output of aten.to to address silent behavior divergence from eager mode. But in some cases, this is too strict because we only want to ban mutation if the output has an alias to it.

Here we relax it to: you can mutate the result of aten.to, but you cannot use it for any computation further.

Under export, aten.to() should always return a FunctionalTensor that aliases its input storage

If a tensor X with the above storage (S) is mutated, we need to flip a bit so that we can error IF:
- Any aliases of X are used in any operation (or if they are returned as graph outputs)
- If X is not a graph input, but its storage is part of a graph input





You can:
```
    def forward(self, x):
        y = x + 1
        z = y.to("cpu")
        z.add_(5)
        return x
```


You can't:

```
    def forward(self, x):
        y = x + 1
        z = y.to("cpu")
        z.add_(5)
        return z
```

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_device_to
buck2 run 'fbcode//mode/dev-nosan' //caffe2/test/functorch:test_aotdispatch -- -r test_aot_export_ban_dropout_mut_pre_dispatch
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:torch -- -r test_to$
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_solver_unsupported_sympy_function$
```

Differential Revision: D64768928

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -537,6 +559,29 @@ def unwrap(x):
+                        for arg in args:
+                            if isinstance(
+                                arg, FunctionalTensor
+                            ) and torch._is_functional_tensor(arg.elem):
```
nit: we have pretty strong invariants that FunctionalTensor's inner tensors are always C++ `FunctionalTensor`s. So you can just remove the `_is_functional_tensor(arg.elem)` check

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -537,6 +559,29 @@ def unwrap(x):
                 else:
+                    arg_mutation_counters = []
+
+                    if self.export:
```
another nit: I'd like to keep the `FunctionalTensorMode.__torch_dispatch__` code as readable as possible, and I'm a bit worried about it being dominated by `is_export` branching code. A few things I think will help:

(1) instead of a separate branch on `self.export` both before and after redispatching, you can probably just have all of your checks run after the redispatch. Even if we are supposed to error because we are tracing some code that passes a frozen tensor into an op, there is nothing wrong with executing that op under tracing first before raising the error

(2) let's move most of the logic for manipulating those private dictionaries into a helper method

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -537,6 +559,29 @@ def unwrap(x):
+                                if (
+                                    arg_group_id in self._mutated_partial_frozen_storage
+                                ):  # type: ignore[attr-defined]
+                                    raise RuntimeError(
```
hmm, how does this code ensure that you do **not** error on code like the below?
```
def f(x):
    y = x + 1
    z = y.to(...)
    z.mul_(2)
    return z
```

I was expected that once you mark a storage as "frozen but mutated", you also remember the specific tensor that was mutated (z in the above example), so you know that z itself is still safe to read from

yushangdi replied:
I see. I'll update the PR to error on this case. I didn't consider this case before, thanks for the catch!

yushangdi replied:
@bdhirsh If we allow this case, I guess we also allow the following case?

```
            def forward(self, x):
                y = x + 1
                z = y.to(torch.int32)
                z.mul_(2)
                z.add_(2)
                return z
```

Technically, `z.add_(2)` is using `z` in a computation. Can we say that we always allow using mutated storage in computation (or return) if the only tensor that mutates this storage is the tensor used in the computation itself?

yushangdi replied:
@bdhirsh I tried to implement this, but now I think it might not be possible in the new training IR.

Given this module:

```python
        class Module(torch.nn.Module):
            def forward(self, x):
                z = x + 1
                y = z.to("cpu")
                z.add_(1)
                return y
```

The training IR graph looks like this:

```
graph():
    %x : [num_users=1] = placeholder[target=x]
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%x, 1), kwargs = {})
    %to : [num_users=1] = call_function[target=torch.ops.aten.to.dtype_layout](args = (%add,), kwargs = {dtype: torch.int64, layout: torch.strided, device: cpu})
    %add_ : [num_users=1] = call_function[target=torch.ops.aten.add_.Tensor](args = (%to, 1), kwargs = {})
    return (add_,)
```

Specifically, originally we are doing z.add_(1), but the training IR graph has add_ on the result of to instead of the input. I think this is expected behavior. So we don't really know if we are returning `y` or `z`....

To keep the behavior of export consistent between all modes, maybe we just don't allow this case?

bdhirsh replied:
@yushangdi oh the example you put above is a bit different: you are mutation `z`, but returning `y`. I agree that we need to ban that behavior (because it is semantically different to the user, depending on whether z and y are aliased).

In the example I gave higher up though, we mutate `z`, and also return `z`. This is ok because we never read from `y`, so it doesn't actually matter if y and z properly aliased or not.

yushangdi replied:
> @yushangdi oh the example you put above is a bit different: you are mutation `z`, but returning `y`. I agree that we need to ban that behavior (because it is semantically different to the user, depending on whether z and y are aliased).
> 
> In the example I gave higher up though, we mutate `z`, and also return `z`. This is ok because we never read from `y`, so it doesn't actually matter if y and z properly aliased or not.

@bdhirsh Yes, the module code is different, but if you look at the fx.graph generated, it's the same as the example you put above. My point is we cannot distinguish the two examples after they are converted to an fx.Graph by export_for_training.

yushangdi replied:
@tugsbayasgalan @bdhirsh 

What do you mean by `y is x`? Do you mean they are the same FunctionalTensor? 
If we do this, wouldn't we error on `z.mul_(2)` in both examples?

> (1) agree to ban both of the above examples (@tugsbayasgalan - my understanding is that some users hit the pessimistic error from example 1 above, so it would be preferable to not error on it?)

If we can't fix the examples above, I'm also wondering if it's worth to land this PR? With this PR, people can mutate the result of aten.to(), but they cannot use the result in any computation, seems like it's still quite restrictive and we had to add a bunch of code. If people are mostly hitting the errors in this example, that seems like landing this PR wouldn't benefit them...

bdhirsh replied:
> What do you mean by y is x? Do you mean they are the same FunctionalTensor?

yep

> If we do this, wouldn't we error on z.mul_(2) in both examples?

and yep (although - we already error on both of those examples today, as you pointed out)

yushangdi replied:
> and yep (although - we already error on both of those examples today, as you pointed out)

I see. So you are saying, when `y` and `x` are the same FunctionalTensor, there's nothing we can do. However, if `y` and `x` are not the same FunctionalTensor, we can further relax it there and allow this case?

bdhirsh replied:
yep!

tugsbayasgalan replied:
@bdhirsh do you have the example where user runs into this issue? I thought their issue was lot simpler haha... If the issue they ran into is indeed simpler, it seems ok to ban both.

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -537,6 +559,29 @@ def unwrap(x):
+                                arg_mutation_counter = torch._mutation_counter(arg.elem)  # type: ignore[attr-defined]
+                                arg_mutation_counters.append(arg_mutation_counter)
+                            else:
+                                arg_mutation_counters.append(0)
```
nit: just make this a dict from `arg_index -> mutation_counter`  instead of putting in zeros for non-tensor args

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -346,6 +357,17 @@ def __init__(self, pre_dispatch=False, export=False, _allow_token_discovery=Fals
         ] = weakref.WeakKeyDictionary()
 
+        # Used for export only.
+        # Map from FunctionalTensor's storage to a group id.
```
Just a nit, but this description doesn't tell me much (what is a group id?).

What do you think of something like this?
```
During export, we want to treat all potentially-aliased tensors that come from `aten.to` as part of a single "aten.to group",
so that we can properly raise an error when any members of the group are mutated.
Whenever a tensor (or its alias) participates as an input/output to `aten.to`, we assign its "aten.to group" a unique id.
```

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -547,16 +592,51 @@ def unwrap(x):
+                                        )
+
+                        # TODO: allow it in draft export mode and attach to FailureReport
+                        if func == torch.ops.aten._to_copy.default:
```
Since we are only ever doing this logic for `aten._to_copy`... do we get the behavior that we want in the case that `.to()` becomes a no-op?

I think it would be good to have a test that we error for e.g. this:
```
def f(x):
    y = x + 1
    # fp32.to(fp32) becomes a no-op
    z = y.to(dtype=torch.float32)
    z.mul_(2)
    # this should error
    return y
    
x = torch.randn(4, dtype=torch.float32)
gm = export(f, (x,))
```

yushangdi replied:
I think this test case is a no-op? I can include a test for your case as well. 


```
        class Module(torch.nn.Module):
            def forward(self, x):
                y = x.to("cpu")
                y.add_(1)
                l = x + 1
                return l

        inps = (torch.tensor(1, device="cpu"),)
```


My understanding is even if `.to()` is a no-op, it still get's traced, and it will be re-dispatched as `aten._to_copy`. here: https://github.com/pytorch/pytorch/blob/80761c62f47df55d6f7c231d23c5fcacc3c0a7f0/torch/_subclasses/functional_tensor.py#L401--L413

bdhirsh replied:
If `x.to(...)` is a no-op, then it will just return `x` directly without dispatching to `_to_copy` (here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorConversions.cpp#L429)

bdhirsh replied:
why is the test above a no-op?

yushangdi replied:
> why is the test above a no-op?

I thought it's a no-op because the input is already in "cpu", so `.to("cpu")` is a no-op. Maybe I'm wrong?

bdhirsh replied:
oh yep, the `.to()` call becomes a no-op when running that code. But I would still expected us to error when attempting to export that code: depending on whether or not you pass a cpu or a cuda tensor to that function, you will get different results.

yushangdi replied:
> oh yep, the `.to()` call becomes a no-op when running that code. But I would still expected us to error when attempting to export that code: depending on whether or not you pass a cpu or a cuda tensor to that function, you will get different results.

sounds good. Yes, we do error when attempting to export that code. 

> If x.to(...) is a no-op, then it will just return x directly without dispatching to _to_copy (here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorConversions.cpp#L429)

We still dispatch to _to_copy because we have a custom `to()` implemented for FunctionalTensor:

https://github.com/pytorch/pytorch/blob/80761c62f47df55d6f7c231d23c5fcacc3c0a7f0/torch/_subclasses/functional_tensor.py#L291

bdhirsh replied:
oh fwiw @tugsbayasgalan - this won't catch any `tensor.to()` calls that happen from C++

tugsbayasgalan replied:
I guess in the long term this is fine because the recommended way to go inference IR is via training IR where we would have materialized aten.to in the first place.

bdhirsh replied:
Just to give a counterexample: you can have a CIA op that shows up directly in training IR, that has a C++ implementation that calls `TensorImpl::to()`

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -547,16 +592,51 @@ def unwrap(x):
+
+                            self._partial_frozen_storage[storage] = group_id
+
+                        # We don't allow any mutation on result of dropout
```
Can you include the reasoning in the comment? My understanding is just that we aren't as worried about people running into this in the `dropout` case, and so we don't want to bother implementing it (although to be fair, once you've implemented this logic for `aten.to()` it should be pretty easy to include dropout)

yushangdi replied:
@tugsbayasgalan Do we want to do the same thing for dropout as well? I remember we discussed this briefly before and decided not to do it for now, but I forgot why. I think we can give dropout the same treatment?

tugsbayasgalan replied:
Hmm i think it was just to not increase the blast radius! We can do it as follow up or in this PR, whatever you prefer.

bdhirsh replied:
sg, I think either is fine (mostly just calling out that the comment makes it unclear to readers whether this is something fundamental or something we want to fix later)

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -301,6 +307,11 @@ def inner(*flat_args):
+
+                if is_export and isinstance(o, FunctionalTensor):
+                    o_group_id = mode._partial_frozen_storage.get(curr_storage, None)
+                    if o_group_id in mode._mutated_partial_frozen_storage:
```
nit: I'd write a helper method in `FunctionalTensorMode` that accepts a storage and returns if it was "frozen but also mutated", instead of poking at the private fields from AOTAutograd

tugsbayasgalan commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -301,6 +307,11 @@ def inner(*flat_args):
+                if is_export and isinstance(o, FunctionalTensor):
+                    o_group_id = mode._partial_frozen_storage.get(curr_storage, None)
+                    if o_group_id in mode._mutated_partial_frozen_storage:
+                        raise RuntimeError("Cannot return on mutated frozen tensor")
```
Maybe more descriptive error message would be good here: "We can return output tensor that has been mutated after aten.to method. Please consider adding .clone() to work around this issue"

tugsbayasgalan commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -537,6 +559,29 @@ def unwrap(x):
+                                    arg_group_id in self._mutated_partial_frozen_storage
+                                ):  # type: ignore[attr-defined]
+                                    raise RuntimeError(
+                                        "Cannot do computation on mutated frozen tensor"
```
Can you add more detail to the error message? Like: "it is unsafe to any computation on a tensor that is mutated as a result of maybe aliasing op, please use .clone() to operate on a fresh tensor"

tugsbayasgalan commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -547,16 +592,51 @@ def unwrap(x):
+                    if self.export:
+                        # check all args. If any of them is an alias to the result of aten.to and is mutated,
+                        # we mark it in _mutated_partial_frozen_storage.
+                        for arg_id, arg in enumerate(args):
```
Could we make it a helper function that abstracts away the export specific logic?

yushangdi committed:
Relax mutation to aten.to in export (#138606)

Summary:

Today, we force aten.to to always copy under the hood and ban the mutation on the output of aten.to to address silent behavior divergence from eager mode. But in some cases, this is too strict because we only want to ban mutation if the output has an alias to it.

Here we relax it to: you can mutate the result of aten.to, but you cannot use it for any computation further.

Under export, aten.to() should always return a FunctionalTensor that aliases its input storage

If a tensor X with the above storage (S) is mutated, we need to flip a bit so that we can error IF:
- Any aliases of X are used in any operation (or if they are returned as graph outputs)
- If X is not a graph input, but its storage is part of a graph input





You can:
```
    def forward(self, x):
        y = x + 1
        z = y.to("cpu")
        z.add_(5)
        return x
```


You can't:

```
    def forward(self, x):
        y = x + 1
        z = y.to("cpu")
        z.add_(5)
        return z
```

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_device_to
buck2 run 'fbcode//mode/dev-nosan' //caffe2/test/functorch:test_aotdispatch -- -r test_aot_export_ban_dropout_mut_pre_dispatch
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:torch -- -r test_to$
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_solver_unsupported_sympy_function
```

Differential Revision: D64768928

bhack commented:
Any update on this?

yushangdi commented:
> Any update on this?

@bhack Unfortunately this approach doesn't work, because of the discussion above. We are trying to do it another way.
